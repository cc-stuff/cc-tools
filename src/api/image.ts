import * as path from 'path';
import * as fs from 'fs-extra';
import {formatFileSize} from "../common/formatFileSize";
import {assertIsString} from "../common/assertIsString";
import {resolvePath} from "../common/resolvePath";

interface FSNodeAttributes {
	creation: number;
	modification: number;
}

interface FSNode {
	attributes: FSNodeAttributes;
}

interface FSFile extends FSNode {
	b64: string;
}

interface FSDirectory extends FSNode {
	children: string[];
}

interface Image {
	[key: string]: FSFile | FSDirectory;
}

function readDir(rootDir: string, dir: string, outImage: Image): string[] {
	const files = fs.readdirSync(path.resolve(rootDir, dir));

	for (const fileName of files) {
		const relFileName = `${dir}/${fileName}`;
		const absFileName = path.resolve(rootDir, relFileName);
		const fileKey = `computer[0].files[${relFileName.slice(2)}]`;

		const fileStats = fs.statSync(absFileName);

		const attributes = {
			creation: Date.now(),
			modification: Date.now(),
		}

		if (fileStats.isDirectory()) {
			const childFiles = readDir(rootDir, relFileName, outImage);

			outImage[fileKey] = {
				attributes,
				children: childFiles,
			}
		} else {
			outImage[fileKey] = {
				attributes,
				b64: fs.readFileSync(absFileName).toString("base64"),
			}
		}
	}

	return files;
}

function flattenImage(image: Image): [string, string][] {
	const result = [];

	for (const key of Object.keys(image)) {
		const entry = image[key];

		result.push([`${key}.attributes`, JSON.stringify(entry.attributes)]);

		if ((entry as FSDirectory).children) {
			result.push([`${key}.children`, JSON.stringify((entry as FSDirectory).children)]);
		} else {
			result.push([`${key}.b64`, JSON.stringify((entry as FSFile).b64)]);
		}
	}

	return result;
}

function writeImage(flat: [string, string][], outDir: string): number {
	let extractedCount = 0;

	for (const entry of flat) {
		const [key, value] = entry;

		const fileName = /^computer\[0]\.files\[(.*)]/.exec(key)[1];
		const fileNameAbs = resolvePath(fileName, outDir);

		// Forcing the path
		fs.mkdirpSync(path.dirname(fileNameAbs));

		// Creating a file
		if (key.endsWith(".b64")) {
			console.log(`Extracting ${fileNameAbs}`);

			fs.writeFileSync(fileNameAbs, Buffer.from(value, "base64").toString("utf-8"));
			extractedCount++;
		}
	}

	return extractedCount;
}

export interface PackImageOptions {
	args?: {
		folder: string;
		output: string;
		noEmit?: boolean;
	}
}

export function packImage(options: PackImageOptions) {
	const {folder, output} = options.args;

	assertIsString(folder);

	const rootDir = resolvePath(folder);

	if (typeof output === "string" && fs.existsSync(output)) {
		fs.unlinkSync(output);
	}

	const image: Image = {};

	image["computer[0].files[]"] = {
		attributes: {
			creation: 0,
			modification: 0,
		},
		children: readDir(rootDir, ".", image),
	}

	const imageOut = JSON.stringify(flattenImage(image));

	if (!options.args.noEmit) {
		assertIsString(output);

		fs.writeFileSync(output, imageOut, "utf-8");

		console.log('Out image: ', formatFileSize(fs.statSync(output).size));
	}

	return imageOut
}

export interface UnpackImageOptions {
	args?: {
		fs?: string;
		image?: string;
		output: string;
	}
}

export function unpackImage(options: UnpackImageOptions) {
	const {fs: imageFile, image, output} = options.args;

	assertIsString(output);

	const outputAbs = resolvePath(output);

	const rawImage = typeof imageFile === "string"
		? fs.readFileSync(resolvePath(imageFile), "utf-8")
		: image;

	const flatImage: [string, string][] = JSON.parse(rawImage);

	const extractedCount = writeImage(flatImage, outputAbs);

	console.log('Extracted files: ', extractedCount);
}
