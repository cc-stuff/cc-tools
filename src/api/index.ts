export {BundleProjectOptions, bundleProject} from "./project";
export {PackImageOptions, packImage} from "./image";
export {RunOptions, ComputerCraftInstance, run} from "./runner";

export interface UnpackImageOptions {
	args?: {
		image: string;
		output: string;
	}
}

export function unpackImage(options: UnpackImageOptions) {
}

export interface TestOptions {
	args?: {
		project: string;
		testsGrep?: string;
		suitsGrep?: string;
	}
}

export function test(options: TestOptions) {
}
