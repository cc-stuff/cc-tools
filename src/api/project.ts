import * as path from 'path';
import * as fs from 'fs';
import {formatFileSize} from "../common/formatFileSize";
import {assertIsString} from "../common/assertIsString";
import {resolvePath} from "../common/resolvePath";
import {MacrosCompiler} from "./macros/MacrosCompiler";

interface CCBundleConfig {
	entry: string;
	output: string;
	references?: Record<string, string>;
}

interface Project {
	config: CCBundleConfig;
	rootDir: string;
	files: Record<string, File>;
	bundleStream: BundleStream;
	macrosCompiler: MacrosCompiler;
}

interface File {
	source: string;
	path: string;
	moduleName: string;
}

const bundleTemplate = (body, entryModuleName) => `
local __files__ = {}

local function __require__(sName)
	local moduleInfo = __files__[sName]

	if (not moduleInfo) then
		error("Module not found: " .. tostring(sName))
	end

	return moduleInfo[2]
end

local function __define__(nLine, sName, fImpl)
	local function strSplit(sStr, sDelimiter)
		local tResult = {}
		local nFrom = 1
		local nDelimFrom, nDelimTo = string.find(sStr, sDelimiter, nFrom)

		while nDelimFrom do
			table.insert(tResult, string.sub(sStr, nFrom, nDelimFrom - 1))
			nFrom = nDelimTo + 1
			nDelimFrom, nDelimTo = string.find(sStr, sDelimiter, nFrom)
		end

		table.insert(tResult, string.sub(sStr, nFrom))

		return tResult
	end

	local function strTrim(s)
		return string.match(s, "^%s*(.-)%s*$")
	end

	local function findTrueSource(nLine)
		local sModule
		local nStart = 0

		for sKey, tModuleInfo in pairs(__files__) do
			if (tModuleInfo.line <= nLine and tModuleInfo.line > nStart) then
				sModule = sKey
				nStart = tModuleInfo.line
			end
		end

		return sModule or "(unknown)", nLine - nStart
	end

	local bStatus, aModule = xpcall(fImpl, function(sError)
		local sStack = debug.traceback()

		-- TODO: Optimize performance
		local tStack = strSplit(sStack, "\\n")
		local sResultError = "Uncaught error: " .. tostring(sError) .. "\\n in module " .. sName .. ":\\n"
		local sCurrentFile

		for i = 1, #tStack do
			local sLine = strTrim(tStack[i])
			local sFile, sLineNumber = table.unpack(strSplit(sLine, ":"))

			if (sCurrentFile == nil and type(sLineNumber) == "number") then
				sCurrentFile = sFile
			end

			if (sFile == sCurrentFile) then
				local sTrueFile, nTrueLine = findTrueSource(sLineNumber)

				sLine = "\\(" .. sTrueFile .. ":" .. tostring(nTrueLine) .. ") " .. sLine
			end

			if (i < 10) then
				sResultError = sResultError .. "\\n" .. sLine
			end
		end

		return sResultError
	end)

	if (bStatus) then
		__files__[sName] = {nLine, aModule}
	else
		error(aModule)
	end
end

${body}

__require__('${entryModuleName}')
`;

const defineTemplate = (line: number, name: string, content: string) => `
__define__(${line}, '${name}', function ()
${content}
end)
`;

class BundleStream {
	private _line: number = 0;
	private _chunks: string[] = [];

	getLine(): number {
		return this._line;
	}

	write(body: string) {
		this._chunks.push(body);
		this._line += (body.match(/\n/g) || []).length;
	}

	writeHeader() {
		const [body] = bundleTemplate("%__body__%", "").split("%__body__%")
		this.write(body);
	}

	writeFooter(entryModuleName: string) {
		const [_, body] = bundleTemplate("%__body__%", entryModuleName).split("%__body__%")
		this.write(body);
	}

	writeModule(name: string, content: string) {
		// TODO: Fix magic number
		this.write(defineTemplate(this.getLine() + 3, name, content));
	}

	toString(): string {
		return this._chunks.join("");
	}
}

const requireTemplate = name => `__require__('${name}')`;

function requireFile(project: Project, parent: File, relativeFileName: string): File {
	const absPath = path.resolve(path.dirname(parent.path), relativeFileName) +
		(relativeFileName.endsWith('.lua') ? '' : '.lua');

	if (project.files[absPath]) {
		return project.files[absPath];
	}

	const file: File = {
		source: fs.readFileSync(absPath, 'utf-8'),
		path: absPath,
		moduleName: path
			.relative(project.rootDir, absPath)
			.replace(/\\/g, "/"),
	}

	file.source = project.macrosCompiler.compile(file.source);

	expandRequires(file, project);

	project.bundleStream.writeModule(file.moduleName, file.source);
	project.files[absPath] = file;

	return file;
}

function expandRequires(file: File, project: Project) {
	const regexp = /require\s+['"](.*)['"]/gmi;
	const result = [];
	let matches;
	let lastIndex = 0;

	while ((matches = regexp.exec(file.source)) !== null) {
		const requiredFile = requireFile(project, file, matches[1]);

		result.push(file.source.substring(lastIndex, matches.index));
		result.push(requireTemplate(requiredFile.moduleName));

		lastIndex = matches.index + matches[0].length;
	}

	result.push(file.source.substring(lastIndex));

	file.source = result.join('');
}

export interface BundleProjectOptions {
	projectFile?: string;
	args?: {
		entry: string;
		output?: string;
		noEmit?: boolean;
	}
}

export function bundleProject(options: BundleProjectOptions) {
	const {projectFile} = options;
	const {entry: entryArg, output: outputArg, noEmit} = options.args ?? {};

	// Creating project
	const project: Project = {
		config: {
			entry: "",
			output: "",
		},
		rootDir: process.cwd(),
		files: {},
		bundleStream: new BundleStream(),
		macrosCompiler: new MacrosCompiler(),
	}

	// Loading the config from the file
	if (typeof projectFile === "string") {
		const projectFileAbs = resolvePath(projectFile);
		const projectRoot = path.dirname(projectFileAbs);

		project.config = JSON.parse(fs.readFileSync(projectFileAbs, "utf-8"));
		project.config.entry = resolvePath(project.config.entry, projectRoot);
		project.config.output = resolvePath(project.config.output, projectRoot);
	} else {
		assertIsString(entryArg);
		project.config.entry = resolvePath(entryArg);

		if (!noEmit) {
			assertIsString(outputArg);

			project.config.output = resolvePath(outputArg);
		}
	}

	project.rootDir = path.dirname(project.config.entry);

	// Header stub file
	const headerFile: File = {
		source: '',
		path: path.resolve(project.rootDir, './__header__'),
		moduleName: '',
	}

	// Writing bundle header
	project.bundleStream.writeHeader();

	// Requiring entry file
	const entryFile = requireFile(project, headerFile, path.basename(project.config.entry));

	// Writing bundle footer
	project.bundleStream.writeFooter(entryFile.moduleName);

	// Validate that it's eiter noEmit or references but never both
	if (noEmit && project.config.references) {
		throw new Error("noEmit can apply only to projects without references")
	}

	const bundleSource = project.bundleStream.toString();

	// Writing bundle to file
	if (!noEmit) {
		fs.writeFileSync(project.config.output, bundleSource, 'utf-8');

		// Bundle size info
		console.log(
			`[${path.basename(project.config.output)}] Out bundle: `,
			formatFileSize(fs.statSync(project.config.output).size),
		);
	}

	// Bundling child projects
	if (project.config.references) {
		for (let reference of Object.values(project.config.references)) {
			reference = resolvePath(reference, project.rootDir);

			bundleProject({projectFile: reference});
		}
	}

	return bundleSource;
}
