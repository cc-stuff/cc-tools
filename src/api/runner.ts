import {Browser, chromium, Page} from "playwright";
import * as path from 'path';
import * as fs from 'fs';
import {createServer} from "http-server";
import * as http from "http";
import {resolvePath} from "../common/resolvePath";
import {EventEmitter} from "events";
import {packImage, unpackImage} from "./image";

declare const __LS_ITEMS__: [string, string][];

interface Context {
	options: RunOptions;
	browserPromise?: Promise<Browser>;
	imageFileAbs?: string;
	image?: string;
	server?: http.Server;
}

function runChromium(context: Context): Promise<Browser> {
	if (!context.browserPromise) {
		context.browserPromise = chromium.launch({
			headless: false,
		});
	}

	return context.browserPromise;
}

export interface ComputerCraftInstance extends EventEmitter {
	readonly page: Page;

	stop(): Promise<void>;
	waitForStatus(status: "started" | "ended"): Promise<void>;
	getRawImage(): string;
}

async function initializePage(context: Context): Promise<ComputerCraftInstance> {
	const imageSrc = typeof context.imageFileAbs === "string"
		? fs.readFileSync(context.imageFileAbs, "utf-8")
		: context.image;

	const image = JSON.parse(imageSrc);

	const browser = await runChromium(context);

	const page = await browser.newPage();

	const initScript = "(" + function () {
		function patchStartupFile(originalFile?: string): string {
			return btoa(`
cctools = {}

cctools.start = function()
	local writeCommand = function(sCommand)
		local file = fs.open(".ccstatus", "w")
		file.write(sCommand)
		file.close()
	end

	cctools.finish = function()
		os.startTimer(1)
		os.pullEvent()
		writeCommand("ended")
	end

	writeCommand("started")
end

cctools.start()

${originalFile === undefined ? "" : atob(originalFile)}
`);
		}

		const items = __LS_ITEMS__;
		const startupKey = "computer[0].files[startup.lua].b64";
		const statusKey = "computer[0].files[.ccstatus].b64";

		localStorage.clear();

		for (const [key, value] of items) {
			let parsedValue = key.endsWith(".b64")
				? JSON.parse(value)
				: value;

			// Patching startup file
			if (key === startupKey) {
				parsedValue = patchStartupFile(parsedValue);
			}

			localStorage.setItem(key, parsedValue);
		}

		// Hook to clear status
		(window as any).__clearStatus__ = () => {
			const statusElement = document.getElementById("ccstatus");

			if (statusElement) {
				statusElement.parentNode.removeChild(statusElement);
			}
		}

		const originalSetItem = localStorage.setItem;

		// Patching localStorage.set to monitor ComputerCraft FS changes
		localStorage.setItem = function (key: string, value: any) {
			// Displaying status in DOM so that Node process can pick it up
			if (key === statusKey) {
				(window as any).__clearStatus__();

				const statusElement = document.createElement("DIV");
				statusElement.id = "ccstatus";
				statusElement.style.display = "none";
				statusElement.innerText = atob(String(value));

				document.body.appendChild(statusElement);

				console.log("status changed", statusElement.innerText);
			}

			originalSetItem.apply(localStorage, arguments);
		}
	}.toString().replace("__LS_ITEMS__", JSON.stringify(image)) + ")()";

	await page.addInitScript(initScript);

	const instance: ComputerCraftInstance = new EventEmitter() as ComputerCraftInstance;
	const server = await startServer(context);
	let running = true;
	let outFs;

	(instance as any).page = page;

	// Method to get output contents of the file system
	instance.getRawImage = () => {
		return outFs;
	}

	// Method to stop this instance
	instance.stop = async () => {
		// Grab the contents of fiel system
		outFs = await page.evaluate(function() {
			const entries: [string, string][] = [];

			for (let i = 0; i < localStorage.length; i++) {
				const key = localStorage.key(i);

				entries.push([key, localStorage.getItem(key)]);
			}

			return JSON.stringify(entries);
		});

		running = false;
		instance.emit("end");

		await Promise.all([
			// Closing the page
			page.close(),
			// Closing the server
			new Promise(resolve => server.close(resolve)),
		]);

		// Closing the browser
		await browser.close();
	}

	// Method to wait for status to come
	instance.waitForStatus = (status: string) => {
		return new Promise(resolve => {
			const listener = (newStatus: string) => {
				if (newStatus === status) {
					instance.off("status", listener);
					resolve();
				}
			}

			instance.on("status", listener);
		});
	}

	(async () => {
		while (running) {
			try {
				const ccStatus = await page.waitForSelector("#ccstatus", {state: "attached"});
				const status = await ccStatus.innerText();

				await page.evaluate("__clearStatus__()")

				instance.emit("status", status);
			} catch (error) {
				console.error(error);
				return;
			}
		}
	})().then()

	return instance;
}

async function startServer(context: Context): Promise<http.Server> {
	if (context.server) {
		return context.server;
	}

	const root = path.resolve(__dirname, "../../vendor/copy-cat");

	const server = createServer({
		root,
	});

	context.server = server;

	server.listen(context.options.args.port);

	const serverImpl: http.Server = typeof (server as any).server?.close === "function"
		? (server as any).server
		: server;

	server.close = function() {
		console.log("Server stopped gracefully");
		return serverImpl.close.apply(serverImpl, arguments);
	}

	console.log(`Copy-cat root is ${root}`);
	console.log(`Listening on http://localhost:${context.options.args.port}`);

	return server;
}

async function main(context: Context): Promise<ComputerCraftInstance> {
	let instance = await initializePage(context);

	await instance.page.goto(`http://localhost:${context.options.args.port}`);

	if (context.options.args.watch) {
		fs.watchFile(context.imageFileAbs, async () => {
			await instance.page.close();
			instance = await initializePage(context);

			await instance.page.goto(`http://localhost:${context.options.args.port}`);
		});
	}

	return instance;
}

export interface RunOptions {
	args?: {
		fs?: string;
		folder?: string;
		port: number;
		watch?: boolean;
	}
}

export function run(options: RunOptions): Promise<ComputerCraftInstance> {
	const {fs: imageFile, folder} = options.args;

	if (typeof imageFile === "string") {
		const imageFileAbs = resolvePath(imageFile);

		return main({
			imageFileAbs,
			options,
		});
	} else {
		const folderAbs = resolvePath(folder);

		const image = packImage({
			args: {
				folder: folderAbs,
				output: undefined,
				noEmit: true,
			}
		});

		return main({
			image,
			options,
		}).then(instance => {
			instance.on("end", () => {
				const imageRaw = instance.getRawImage();

				// Unpacking the image after the test is sun
				unpackImage({
					args: {
						image: imageRaw,
						output: folderAbs,
					}
				});
			});

			return instance;
		})
	}
}
