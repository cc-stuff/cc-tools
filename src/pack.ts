import {argv} from 'yargs';
import {assertIsString} from "./common/assertIsString";
import {packImage} from "./api";

const {folder, output} = argv;

assertIsString(folder);
assertIsString(output);

packImage({
	args: {
		folder,
		output,
	}
});
