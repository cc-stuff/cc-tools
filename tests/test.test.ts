import {test as runTest} from "../src/api";

describe("test", () => {

	test("run all tests", async () => {
		const outputLines = [];
		let error;

		try {
			await runTest({
				outputLines,
				args: {
					project: "./tests/assets/test1/project.json",
				}
			});
		} catch (err) {
			error = err;
			console.error(err);
		}

		expect(outputLines).toMatchSnapshot();
		expect(error).toMatchSnapshot();
	});

	test("run all tests that fail", async () => {
		const outputLines = [];
		let error;

		try {
			await runTest({
				outputLines,
				args: {
					project: "./tests/assets/test2/project.json",
				}
			});
		} catch (err) {
			error = err;
			console.error(err);
		}

		expect(outputLines).toMatchSnapshot();
		expect(error).toMatchSnapshot();
	});

	test("run tests with hooks", async () => {
		const outputLines = [];
		let error;

		try {
			await runTest({
				outputLines,
				args: {
					project: "./tests/assets/test3/project.json",
				}
			});
		} catch (err) {
			error = err;
			console.error(err);
		}

		expect(outputLines).toMatchSnapshot();
		expect(error).toMatchSnapshot();
	});

});
