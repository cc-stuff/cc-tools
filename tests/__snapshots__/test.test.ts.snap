// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test run all tests 1`] = `
Array [
  "[42m[37mPASS[39m[49m [32mBasic[39m",
  "[42m[37mPASS[39m[49m [32mTable equality[39m",
  "",
  "Total tests: 11",
  "Failed tests: 0",
]
`;

exports[`test run all tests 2`] = `undefined`;

exports[`test run all tests that fail 1`] = `
Array [
  "[41m[37mFAIL[39m[49m [31mFailures[39m",
  "[1m[31m    â€¢ not to be falsy[39m[22m",
  "[31mnot expecting false to be falsy[39m",
  "",
  "[1m[31m    â€¢ not to be nil[39m[22m",
  "[31mnot expecting nil to be nil[39m",
  "",
  "[1m[31m    â€¢ tables equal[39m[22m",
  "[31mexpecting values to match[39m",
  "",
  "{	\\"number\\" = 2	\\"subtable\\" = {		\\"a\\" = true",
  "[31m[-] 	}[39m",
  "[32m[+] 		\\"b\\" = false[39m",
  "",
  "[31m[-] 	\\"text\\" = \\"2\\"[39m",
  "[32m[+] 	}[39m",
  "[32m[+] [39m",
  "[32m[+] 	\\"text\\" = \\"3\\"[39m",
  "}",
  "[1m[31m    â€¢ to be[39m[22m",
  "[31mexpecting 2 to be 4[39m",
  "",
  "[1m[31m    â€¢ to be falsy[39m[22m",
  "[31mexpecting true to be falsy[39m",
  "",
  "[1m[31m    â€¢ to be nil[39m[22m",
  "[31mexpecting true to be nil[39m",
  "",
  "[1m[31m    â€¢ to be truthy[39m[22m",
  "[31mexpecting false to be truthy[39m",
  "",
  "",
  "Total tests: 7",
  "Failed tests: 7",
]
`;

exports[`test run all tests that fail 2`] = `[Error: 7 tests failed.]`;

exports[`test run tests with hooks 1`] = `
Array [
  "[42m[37mPASS[39m[49m [32mHooks[39m",
  "",
  "Total tests: 1",
  "Failed tests: 0",
]
`;

exports[`test run tests with hooks 2`] = `undefined`;

exports[`test run tests with print 1`] = `
Array [
  "[42m[37mPASS[39m[49m [32mPrint[39m",
  "    â€¢ print() in print toSnapshot array",
  "{",
  "	\\"1\\" = 1",
  "	\\"2\\" = 2",
  "	\\"3\\" = 3",
  "	\\"4\\" = \\"some string\\"}",
  "",
  "    â€¢ print() in print toSnapshot table",
  "{",
  "	\\"func\\" = function: ec1bd8cb",
  "	\\"number\\" = 2",
  "	\\"string\\" = \\"str\\"",
  "	\\"subtable\\" = {",
  "	",
  "	\\"exists\\" = true",
  "	}}",
  "",
  "",
  "Total tests: 2",
  "Failed tests: 0",
]
`;

exports[`test run tests with print 2`] = `undefined`;

exports[`test run tests with traceback 1`] = `
Array [
  "[41m[37mFAIL[39m[49m [31mTraceback in hooks[39m",
  "[1m[31m    â€¢ Traceback in hooks.afterAll[39m[22m",
  "[31mError in hook Traceback in hooks.afterAll: traceback.test.lua.tmp.bundle:36: Custom error in afterAllstack traceback:[39m",
  "[31m	startup.lua:522: in function 'error'[39m",
  "[31m	traceback.test.lua.tmp.bundle:36: in function <traceback.test.lua.tmp.bundle:35>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	[C]: in function 'xpcall'[39m",
  "[31m	startup.lua:514: in local 'runAll'[39m",
  "[31m	startup.lua:592: in field 'run'[39m",
  "[31m	startup.lua:720: in function <startup.lua:713>[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	startup.lua:757: in local 'impl'[39m",
  "[31m	startup.lua:30: in local '__define__'[39m",
  "[31m	startup.lua:734: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	shell.lua:82: in function <shell.lua:62>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	startup.lua:178: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	shell.lua:82: in function <shell.lua:62>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	shell.lua:591: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	multishell.lua:81: in function <multishell.lua:80>[39m",
  "",
  "[1m[31m    â€¢ Traceback in hooks.afterEach[39m[22m",
  "[31mError in hook Traceback in hooks.afterEach: traceback.test.lua.tmp.bundle:32: Custom error in afterEachstack traceback:[39m",
  "[31m	startup.lua:522: in function 'error'[39m",
  "[31m	traceback.test.lua.tmp.bundle:32: in function <traceback.test.lua.tmp.bundle:31>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	[C]: in function 'xpcall'[39m",
  "[31m	startup.lua:514: in local 'runAll'[39m",
  "[31m	startup.lua:586: in field 'run'[39m",
  "[31m	startup.lua:720: in function <startup.lua:713>[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	startup.lua:757: in local 'impl'[39m",
  "[31m	startup.lua:30: in local '__define__'[39m",
  "[31m	startup.lua:734: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	shell.lua:82: in function <shell.lua:62>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	startup.lua:178: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	shell.lua:82: in function <shell.lua:62>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	shell.lua:591: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	multishell.lua:81: in function <multishell.lua:80>[39m",
  "",
  "[1m[31m    â€¢ Traceback in hooks.beforeAll[39m[22m",
  "[31mError in hook Traceback in hooks.beforeAll: traceback.test.lua.tmp.bundle:24: Custom error in beforeAllstack traceback:[39m",
  "[31m	startup.lua:522: in function 'error'[39m",
  "[31m	traceback.test.lua.tmp.bundle:24: in function <traceback.test.lua.tmp.bundle:23>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	[C]: in function 'xpcall'[39m",
  "[31m	startup.lua:514: in local 'runAll'[39m",
  "[31m	startup.lua:576: in field 'run'[39m",
  "[31m	startup.lua:720: in function <startup.lua:713>[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	startup.lua:757: in local 'impl'[39m",
  "[31m	startup.lua:30: in local '__define__'[39m",
  "[31m	startup.lua:734: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	shell.lua:82: in function <shell.lua:62>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	startup.lua:178: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	shell.lua:82: in function <shell.lua:62>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	shell.lua:591: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	multishell.lua:81: in function <multishell.lua:80>[39m",
  "",
  "[1m[31m    â€¢ Traceback in hooks.beforeEach[39m[22m",
  "[31mError in hook Traceback in hooks.beforeEach: traceback.test.lua.tmp.bundle:28: Custom error in beforeEachstack traceback:[39m",
  "[31m	startup.lua:522: in function 'error'[39m",
  "[31m	traceback.test.lua.tmp.bundle:28: in function <traceback.test.lua.tmp.bundle:27>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	[C]: in function 'xpcall'[39m",
  "[31m	startup.lua:514: in local 'runAll'[39m",
  "[31m	startup.lua:582: in field 'run'[39m",
  "[31m	startup.lua:720: in function <startup.lua:713>[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	startup.lua:757: in local 'impl'[39m",
  "[31m	startup.lua:30: in local '__define__'[39m",
  "[31m	startup.lua:734: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	shell.lua:82: in function <shell.lua:62>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	startup.lua:178: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	shell.lua:82: in function <shell.lua:62>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	shell.lua:591: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	multishell.lua:81: in function <multishell.lua:80>[39m",
  "",
  "[41m[37mFAIL[39m[49m [31mTraceback in suite[39m",
  "[1m[31m    â€¢ [39m[22m",
  "[31mError in Traceback in suite: traceback.test.lua.tmp.bundle:17: Custom error in suitestack traceback:[39m",
  "[31m	startup.lua:672: in function 'error'[39m",
  "[31m	traceback.test.lua.tmp.bundle:17: in function <traceback.test.lua.tmp.bundle:15>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	[C]: in function 'xpcall'[39m",
  "[31m	startup.lua:664: in function 'describe'[39m",
  "[31m	traceback.test.lua.tmp.bundle:15: in local 'impl'[39m",
  "[31m	traceback.test.lua.tmp.bundle:9: in local '__define__'[39m",
  "[31m	traceback.test.lua.tmp.bundle:13: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	startup.lua:747: in local 'impl'[39m",
  "[31m	...[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	shell.lua:82: in function <shell.lua:62>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	startup.lua:178: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	shell.lua:82: in function <shell.lua:62>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	shell.lua:591: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	multishell.lua:81: in function <multishell.lua:80>[39m",
  "",
  "[41m[37mFAIL[39m[49m [31mTraceback in test[39m",
  "[1m[31m    â€¢ traceback[39m[22m",
  "[31mError in test Traceback in test.traceback: traceback.test.lua.tmp.bundle:47: Custom error in teststack traceback:[39m",
  "[31m	startup.lua:554: in function 'error'[39m",
  "[31m	traceback.test.lua.tmp.bundle:47: in function <traceback.test.lua.tmp.bundle:46>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	[C]: in function 'xpcall'[39m",
  "[31m	startup.lua:543: in local 'runTest'[39m",
  "[31m	startup.lua:584: in field 'run'[39m",
  "[31m	startup.lua:720: in function <startup.lua:713>[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	startup.lua:757: in local 'impl'[39m",
  "[31m	startup.lua:30: in local '__define__'[39m",
  "[31m	startup.lua:734: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	shell.lua:82: in function <shell.lua:62>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	startup.lua:178: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	shell.lua:82: in function <shell.lua:62>[39m",
  "[31m	(...tail calls...)[39m",
  "[31m	shell.lua:591: in main chunk[39m",
  "[31m	[C]: in function 'pcall'[39m",
  "[31m	bios.lua:528: in field 'run'[39m",
  "[31m	multishell.lua:81: in function <multishell.lua:80>[39m",
  "",
  "",
  "Total tests: 2",
  "Failed tests: 1",
]
`;

exports[`test run tests with traceback 2`] = `[Error: 1 tests failed.]`;
